## **Pandas Essencial**
**Prof. Dr. Samuel Martins (@hisamuka @xavecoding)** <br/>
xavecoding: https://youtube.com/c/xavecoding <br/>

Neste tutorial, vamos aprender o essencial da biblioteca _Pandas_ para manipulação de dados.<br/><br/>

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

### Pacotes usados neste Notebook

In [1]:
# pacotes usados neste notebook
import pandas as pd

<h1>Manipulação de Dados com Pandas</h1><hr/>

<h2>1. Manipulação Básica de Datasets</h2>
<hr/>

**Dataset**: Gas Prices in Brazil: https://www.kaggle.com/matheusfreitag/gas-prices-in-brazil <br/>

Este dataset contém os **registros dos preços médios semanais dos combustíveis do Brasil entre os anos de 2004 e 2019**. <br/>
Cada *amostra (registro/linha)* consiste em um registro de preço aferido para um dado tipo de combustível em uma dada localidade do Brasil. <br/>
Alguns dos principais *atributos* (colunas) do dataset são: 'ESTADO', 'PRODUTO', 'NÚMERO DE POSTOS PESQUISADOS', 'PREÇO MÉDIO REVENDA'.


\* O arquivo disponibilizado no Kaggle está no formato *tsv*. Embora o _pandas_ consiga abrí-lo normalmente, convertemos tal arquivo para o formato *CSV*, que é um dos formatos mais utilizados, e mudamos seu separado para ';' apenas para mostrar algumas opções da função de carregamento.

### 1.1. Importando o Dataset
Para carregar um dataset no formato csv, basta utilizar a função `read_csv` do pandas. Por padrão, ela considera _','_ como separador.

In [3]:
df = pd.read_csv("GasPricesinBrazil_2004-2019.csv")

In [4]:
df.head()

Unnamed: 0;DATA INICIAL;DATA FINAL;REGIÃO;ESTADO;PRODUTO;NÚMERO DE POSTOS PESQUISADOS;UNIDADE DE MEDIDA;PREÇO MÉDIO REVENDA;DESVIO PADRÃO REVENDA;PREÇO MÍNIMO REVENDA;PREÇO MÁXIMO REVENDA;MARGEM MÉDIA REVENDA;COEF DE VARIAÇÃO REVENDA;PREÇO MÉDIO DISTRIBUIÇÃO;DESVIO PADRÃO DISTRIBUIÇÃO;PREÇO MÍNIMO DISTRIBUIÇÃO;PREÇO MÁXIMO DISTRIBUIÇÃO;COEF DE VARIAÇÃO DISTRIBUIÇÃO;MÊS;ANO
0  0;2004-05-09;2004-05-15;CENTRO OESTE;DISTRITO ...                                                                                                                                                                                                                                                                                                                                     
1  1;2004-05-09;2004-05-15;CENTRO OESTE;GOIAS;ETA...                                                                                                                                                                                                                                                                                                                                     
2  2;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...                                                                                                                                                                                                                                                                                                                                     
3  3;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...                                                                                                                                                                                                                                                                                                                                     
4  4;2004-05-09;2004-05-15;NORDESTE;ALAGOAS;ETANO...

O dataset não foi carregado corretamente pois o separador utilizado seu arquivo CSV era ';' e não a ','. <br/>
Vamos então carregá-lo corretamente:

In [5]:
df = pd.read_csv("GasPricesinBrazil_2004-2019.csv", sep=';')
df.head()

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0           0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1           1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2           2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3           3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4           4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  PREÇO MÁXIMO REVENDA  \
0                1.288                  0.016  ...                 1.350   
1                1.162                  0.114  ...                 1.449   
2                1.389                  0.097  ...                 1.760   
3                1.262                  0.070  ...                 1.509   
4                1.181                  0.078  ...                 1.400   

   MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  PREÇO MÉDIO DISTRIBUIÇÃO  \
0                 0.463                    0.012                     0.825   
1                 0.399                    0.098                     0.763   
2                 0.419                    0.070                      0.97   
3                 0.432                    0.055                      0.83   
4                  0.24                    0.066                     0.941   

  DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                       0.11                    0.4201   
1                      0.088                    0.5013   
2                      0.095                    0.5614   
3                      0.119                    0.5991   
4                      0.077                    0.7441   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO  
0                    0.9666                         0.133   5  2004  
1                      1.05                         0.115   5  2004  
2                     1.161                         0.098   5  2004  
3                   1.22242                         0.143   5  2004  
4                    1.0317                         0.082   5  2004  

[5 rows x 21 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106823 entries, 0 to 106822
Data columns (total 21 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Unnamed: 0                     106823 non-null  int64  
 1   DATA INICIAL                   106823 non-null  object 
 2   DATA FINAL                     106823 non-null  object 
 3   REGIÃO                         106823 non-null  object 
 4   ESTADO                         106823 non-null  object 
 5   PRODUTO                        106823 non-null  object 
 6   NÚMERO DE POSTOS PESQUISADOS   106823 non-null  int64  
 7   UNIDADE DE MEDIDA              106823 non-null  object 
 8   PREÇO MÉDIO REVENDA            106823 non-null  float64
 9   DESVIO PADRÃO REVENDA          106823 non-null  float64
 10  PREÇO MÍNIMO REVENDA           106823 non-null  float64
 11  PREÇO MÁXIMO REVENDA           106823 non-null  float64
 12  MARGEM MÉDIA REVENDA          

### 1.2. Exibindo as primeiras linhas do Dataset
A função `.head()` exibe as 5 primeiras linhas do dataset/tabela/Data Frame.

In [7]:
df.head()

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0           0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1           1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2           2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3           3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4           4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  PREÇO MÁXIMO REVENDA  \
0                1.288                  0.016  ...                 1.350   
1                1.162                  0.114  ...                 1.449   
2                1.389                  0.097  ...                 1.760   
3                1.262                  0.070  ...                 1.509   
4                1.181                  0.078  ...                 1.400   

   MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  PREÇO MÉDIO DISTRIBUIÇÃO  \
0                 0.463                    0.012                     0.825   
1                 0.399                    0.098                     0.763   
2                 0.419                    0.070                      0.97   
3                 0.432                    0.055                      0.83   
4                  0.24                    0.066                     0.941   

  DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                       0.11                    0.4201   
1                      0.088                    0.5013   
2                      0.095                    0.5614   
3                      0.119                    0.5991   
4                      0.077                    0.7441   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO  
0                    0.9666                         0.133   5  2004  
1                      1.05                         0.115   5  2004  
2                     1.161                         0.098   5  2004  
3                   1.22242                         0.143   5  2004  
4                    1.0317                         0.082   5  2004  

[5 rows x 21 columns]

In [8]:
list(df.columns)

['Unnamed: 0',
 'DATA INICIAL',
 'DATA FINAL',
 'REGIÃO',
 'ESTADO',
 'PRODUTO',
 'NÚMERO DE POSTOS PESQUISADOS',
 'UNIDADE DE MEDIDA',
 'PREÇO MÉDIO REVENDA',
 'DESVIO PADRÃO REVENDA',
 'PREÇO MÍNIMO REVENDA',
 'PREÇO MÁXIMO REVENDA',
 'MARGEM MÉDIA REVENDA',
 'COEF DE VARIAÇÃO REVENDA',
 'PREÇO MÉDIO DISTRIBUIÇÃO',
 'DESVIO PADRÃO DISTRIBUIÇÃO',
 'PREÇO MÍNIMO DISTRIBUIÇÃO',
 'PREÇO MÁXIMO DISTRIBUIÇÃO',
 'COEF DE VARIAÇÃO DISTRIBUIÇÃO',
 'MÊS',
 'ANO']

In [16]:
df.drop(columns='Unnamed: 0')

DATA INICIAL  DATA FINAL        REGIÃO               ESTADO  \
0        2004-05-09  2004-05-15  CENTRO OESTE     DISTRITO FEDERAL   
1        2004-05-09  2004-05-15  CENTRO OESTE                GOIAS   
2        2004-05-09  2004-05-15  CENTRO OESTE          MATO GROSSO   
3        2004-05-09  2004-05-15  CENTRO OESTE   MATO GROSSO DO SUL   
4        2004-05-09  2004-05-15      NORDESTE              ALAGOAS   
...             ...         ...           ...                  ...   
106818   2019-06-23  2019-06-29      NORDESTE  RIO GRANDE DO NORTE   
106819   2019-06-23  2019-06-29           SUL    RIO GRANDE DO SUL   
106820   2019-06-23  2019-06-29           SUL       SANTA CATARINA   
106821   2019-06-23  2019-06-29       SUDESTE            SAO PAULO   
106822   2019-06-23  2019-06-29      NORDESTE              SERGIPE   

                 PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0       ETANOL HIDRATADO                           127              R$/l   
1       ETANOL HIDRATADO                           387              R$/l   
2       ETANOL HIDRATADO                           192              R$/l   
3       ETANOL HIDRATADO                           162              R$/l   
4       ETANOL HIDRATADO                           103              R$/l   
...                  ...                           ...               ...   
106818               GNV                             7             R$/m3   
106819               GNV                            23             R$/m3   
106820               GNV                            24             R$/m3   
106821               GNV                            52             R$/m3   
106822               GNV                             4             R$/m3   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                     1.288                  0.016                 1.190   
1                     1.162                  0.114                 0.890   
2                     1.389                  0.097                 1.180   
3                     1.262                  0.070                 1.090   
4                     1.181                  0.078                 1.050   
...                     ...                    ...                   ...   
106818                3.574                  0.065                 3.499   
106819                3.401                  0.129                 3.230   
106820                2.912                  0.190                 2.699   
106821                3.020                  0.229                 2.699   
106822                3.697                  0.005                 3.689   

        PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                      1.350                0.463                     0.012   
1                      1.449                0.399                     0.098   
2                      1.760                0.419                     0.070   
3                      1.509                0.432                     0.055   
4                      1.400                 0.24                     0.066   
...                      ...                  ...                       ...   
106818                 3.690                0.818                     0.018   
106819                 3.789                 0.95                     0.038   
106820                 3.499                0.914                     0.065   
106821                 3.490                0.646                     0.076   
106822                 3.699                1.179                     0.001   

       PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                         0.825                       0.11   
1                         0.763                      0.088   
2                          0.97                      0.095   
3                          0.83                      0.119   
4                         0.941                      0.077   
...                

In [14]:
df.head()

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0           0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1           1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2           2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3           3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4           4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  PREÇO MÁXIMO REVENDA  \
0                1.288                  0.016  ...                 1.350   
1                1.162                  0.114  ...                 1.449   
2                1.389                  0.097  ...                 1.760   
3                1.262                  0.070  ...                 1.509   
4                1.181                  0.078  ...                 1.400   

   MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  PREÇO MÉDIO DISTRIBUIÇÃO  \
0                 0.463                    0.012                     0.825   
1                 0.399                    0.098                     0.763   
2                 0.419                    0.070                      0.97   
3                 0.432                    0.055                      0.83   
4                  0.24                    0.066                     0.941   

  DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                       0.11                    0.4201   
1                      0.088                    0.5013   
2                      0.095                    0.5614   
3                      0.119                    0.5991   
4                      0.077                    0.7441   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO  
0                    0.9666                         0.133   5  2004  
1                      1.05                         0.115   5  2004  
2                     1.161                         0.098   5  2004  
3                   1.22242                         0.143   5  2004  
4                    1.0317                         0.082   5  2004  

[5 rows x 21 columns]

### 1.3 Informações do Dataset e Elementos Chave

### 1.3.1 Informações gerais sobre o Dataset

In [27]:
df2 = df[(df.ESTADO == 'ALAGOAS') & (df.PRODUTO == 'ÓLEO DIESEL') & (df.ANO == 2019)]
df2.head()

Unnamed: 0 DATA INICIAL  DATA FINAL    REGIÃO   ESTADO      PRODUTO  \
103096      103096   2019-01-06  2019-01-12  NORDESTE  ALAGOAS  ÓLEO DIESEL   
103246      103246   2019-01-13  2019-01-19  NORDESTE  ALAGOAS  ÓLEO DIESEL   
103396      103396   2019-01-20  2019-01-26  NORDESTE  ALAGOAS  ÓLEO DIESEL   
103546      103546   2019-01-27  2019-02-02  NORDESTE  ALAGOAS  ÓLEO DIESEL   
103696      103696   2019-02-03  2019-02-09  NORDESTE  ALAGOAS  ÓLEO DIESEL   

        NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
103096                            38              R$/l                3.715   
103246                            36              R$/l                3.670   
103396                            39              R$/l                3.657   
103546                            39              R$/l                3.679   
103696                            44              R$/l                3.676   

        DESVIO PADRÃO REVENDA  ...  PREÇO MÁXIMO REVENDA  \
103096                  0.166  ...                 4.089   
103246                  0.185  ...                 4.070   
103396                  0.185  ...                 4.070   
103546                  0.184  ...                 4.165   
103696                  0.190  ...                 4.165   

        MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  \
103096                 0.573                    0.045   
103246                 0.476                    0.050   
103396                 0.446                    0.051   
103546                 0.457                    0.050   
103696                 0.444                    0.052   

        PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
103096                     3.142                      0.122   
103246                     3.194                      0.125   
103396                     3.211                      0.122   
103546                     3.222                      0.118   
103696                     3.232                      0.109   

       PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
103096                      2.99                    3.3741   
103246                      2.99                    3.4042   
103396                      2.99                    3.4175   
103546                      3.07                     3.458   
103696                      3.07                     3.458   

       COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO  
103096                         0.039   1  2019  
103246                         0.039   1  2019  
103396                         0.038   1  2019  
103546                         0.037   1  2019  
103696                         0.034   2  2019  

[5 rows x 21 columns]

In [28]:
df2['ESTADO'] = 'Alagoas'

C:\Users\Douglas Tinoco\AppData\Local\Temp\ipykernel_12780\3795670901.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['ESTADO'] = 'Alagoas'


In [32]:
df[df.ESTADO == 'ALAGOAS']

Unnamed: 0 DATA INICIAL  DATA FINAL    REGIÃO   ESTADO  \
4                4   2004-05-09  2004-05-15  NORDESTE  ALAGOAS   
31              31   2004-05-16  2004-05-22  NORDESTE  ALAGOAS   
58              58   2004-05-23  2004-05-29  NORDESTE  ALAGOAS   
85              85   2004-05-30  2004-06-05  NORDESTE  ALAGOAS   
112            112   2004-06-06  2004-06-12  NORDESTE  ALAGOAS   
...            ...          ...         ...       ...      ...   
106701      106701   2019-06-23  2019-06-29  NORDESTE  ALAGOAS   
106728      106728   2019-06-23  2019-06-29  NORDESTE  ALAGOAS   
106755      106755   2019-06-23  2019-06-29  NORDESTE  ALAGOAS   
106782      106782   2019-06-23  2019-06-29  NORDESTE  ALAGOAS   
106808      106808   2019-06-23  2019-06-29  NORDESTE  ALAGOAS   

                 PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
4       ETANOL HIDRATADO                           103              R$/l   
31      ETANOL HIDRATADO                           106              R$/l   
58      ETANOL HIDRATADO                           103              R$/l   
85      ETANOL HIDRATADO                           105              R$/l   
112     ETANOL HIDRATADO                           103              R$/l   
...                  ...                           ...               ...   
106701       ÓLEO DIESEL                            35              R$/l   
106728   ÓLEO DIESEL S10                            48              R$/l   
106755    GASOLINA COMUM                            57              R$/l   
106782               GLP                            26           R$/13Kg   
106808               GNV                             8             R$/m3   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  PREÇO MÁXIMO REVENDA  \
4                     1.181                  0.078  ...                  1.40   
31                    1.182                  0.069  ...                  1.40   
58                    1.182                  0.076  ...                  1.40   
85                    1.184                  0.083  ...                  1.55   
112                   1.201                  0.060  ...                  1.40   
...                     ...                    ...  ...                   ...   
106701                3.702                  0.158  ...                  4.10   
106728                3.782                  0.147  ...                  4.15   
106755                4.554                  0.231  ...                  5.11   
106782               65.016                  3.515  ...                 73.00   
106808                3.458                  0.085  ...                  3.70   

        MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  \
4                       0.24                    0.066   
31                     0.257                    0.058   
58                     0.252                    0.064   
85                     0.269                    0.070   
112                    0.259                    0.050   
...                      ...                      ...   
106701                 0.448                    0.043   
106728                 0.333                    0.039   
106755                 0.528                    0.051   
106782                16.847                    0.054   
106808                 1.007                    0.025   

        PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
4                          0.941                      0.077   
31                         0.925                      0.065   
58                          0.93                      0.079   
85                         0.915                      0.074   
112                        0.942                      0.083   
...                          ...                        ...   
106701                     3.254                      0.099   
106728                     3.449                       0.08   
106755                     4.026                      0.098   
106782        

A primeira coluna da tabela, chamada _'Unnamed: 0'_, parece não significar nada. Na verdade, ela parece ser os **índices** da tabela, que foram salvos como uma _coluna_.<br/>
Veremos jajá como **remover tal coluna**.

Outro ponto é que, aparentemente, nenhum atributo/coluna possui valores nulos (_null_), uma vez que o número de registros do dataframe e os números de valores _non-null_ é de **106823**. <br/>
Mas, veremos que não é bem assim para esse caso.

### 1.3.2 Data Frame
Todo dataset carregado (dados estruturados) é um `Data Frame`: 'Tabela' bi-dimensional, de tamanho mutável, com dados potencialmente heterogêneos. <br/>

Podemos acessar as **dimensões do Data Frame** (número de linhas x número de colunas) utilizando o atributo `.shape` do Data Frame.

#### **Criando um DataFrame**

Podemos criar um DataFrame a partir de um _dicionário_, onde cada **chave** possui uma **lista de elementos de igual tamanho**.<br/>
As **chaves** representam as **colunas** e **cada um dos valores de sua lista** representa o **valor da linha** correspondente para aquela coluna.

In [ ]:
alunos_df = pd.DataFrame({
    'nome': ['Luke Skywalker', 'Yoda', 'Palpatine'],
    'idade': [16, 1000, 70],
    'peso': [70, 15, 60],
    'eh jedi': [True, True, False]  # o nome das colunas podem ter espaços
})

In [ ]:
alunos_df

#### **VEJA MAIS**
Criando um Data Frame a partir de um dicionário: https://www.geeksforgeeks.org/how-to-create-dataframe-from-dictionary-in-python-pandas/

#### **Renomeando as colunas de um DataFrame**
**===>** O método `DataFrame.columns` retorna uma ";lista" com os **nomes de todas as colunas** do data frame.

<br/>

**===>** Para **renomear colunas** do data frame, utilize o método `DataFrame.rename`, que retorna uma _cópia_ do data frame com as as colunas renomeadas:

Para renomear o _próprio_ data frame em questão, utilize o parâmetro `inplace=True`:

<br/>

**===>** Uma outra forma de **renomear todas as listas** de um data frame é passar uma _lista_ com os novos nomes das colunas para o atributo `DataFrame.columns`:

### 1.3.3 Series

Array uni-dimensional com os dados e rótulos de um eixo.

#### **Criando uma Series**

Podemos criar um DataFrame a partir de uma lista de elementos.

Podemos alterar o **nome dos índices** (veremos melhor jajá) e o **nome da Series** (o que ela representa):

#### **VEJA MAIS**
https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html

### 1.3.4 Atribuindo Dados

#### 1.3.4.1 Atribuindo constantes

#### 1.3.4.2 Atribuindo listas ou series

#### 1.3.4.3 Criando novas colunas
Para **criar uma nova coluna** em um data frame, basta atribuirmos uma _lista/Series de valores_ a uma **nova 'chave'** do data frame. <br/>

**PS:** A _quantidade de valores_ da lista precisa ser **igual** ao _número de linhas/registros_ do data frame.

<br/>

Outro exemplo:

**PS:** Obviamente, a lógica correta em converter o preço dos combustíveis em reais para dólares não é considerar uma taxa de câmbio fixa, uma vez que cada preço foi aferido em um momento diferente.

### 1.3.4 Índices

Todo Data Frame possui **índices**, que não são considerado colunas da tabela. Tais índices são comumente **númericos**, de 0 a num_linhas-1, mas também podem ser **textuais (rótulos/labels)**.

Use `list(data.index)` ou `data.index.to_list()` para converter um RangeIndex para uma python list.

#### **Exemplo de Data Frame com Índices Textuais (labels)**

In [ ]:
pesquisa_de_satisfacao = pd.DataFrame({
    'bom': [50, 21, 100],
    'ruim': [131, 2, 30],
    'pessimo': [30, 20, 1]
}, index=['XboxOne', 'Playstation4', 'Switch'])

In [ ]:
pesquisa_de_satisfacao.head()

### 1.4 Selecionando uma ou mais amostras (Indexação)

#### **==>  Index-based selection (seleção baseada em Índices)**
Mostrando linhas específicas de um DataFrame:

`iloc`: seleciona elementos do Dataframe, baseado em seu **índice (número)** --> row-first, column-second

**Selecionando uma amostra/linha:**

**Selecionando múltiplias amostras/linhas:**

In [ ]:
# e assim por diante!

#### **==>  Label-based selection (seleção baseadas em Rótulos)**

`loc`: seleciona elementos do Dataframe, baseado em seus **rótulos** --> row-first, column-second

### 1.5 Selecionando um ou mais atributos (colunas)

Como o rótulo da coluna 'DATA INICIAL' **possui espaço**, não é possível acessá-la como método: `data.DATA INICIAL`

### 1.6 Removendo um Atributo (Coluna) do Data Frame

Como vimos anteriormente, o atributo 'Unnamed: 0' parece ser um **ruído** em nosso dataset. Desta maneira, vamos eliminá-lo,

### 1.7 Salvando um Data Frame

Para salvarmos um Data Frame para um **arquivo CSV**, basta usarmos o método `.to_csv`. <br/>
Por padrão, esse método **salva os índices da tabela como uma coluna no CSV**.<br/>
Como no geral tais índices são números de 0 a n-1, não há necessidade para isso (veja que removemos anteriormente a coluna 'Unnamed: 0' que foi justamente esse caso).<br/>
Desta forma, utilize o parâmetro: `index=False`.

Por padrão, o método utilizará a ',' como separador das colunas. Caso queira alterá-lo, utilize o parâmetro `sep`.

### 1.8 Seleção Condicional: Filtrando amostras

Durante nossas análise exploratórias, frequentemente filtraremos nossas amostras, a partir de certas **condições**, para fins de análise mais específica. <br/>
Existem algumas maneiras de fazermos tal filtragem. Antes disso, vamos carregar nosso dataset pré-processado que salvamos no item anterior.

#### **Selecionando apenas os preços dos Postos de São Paulo**

##### **==> Alternativa 1: Seleção Condicional (Comparações diretas)**

O código abaixo retorna uma ***Series ('array') de booleans***, com o número de linhas (amostras) do Data Frame, que informa os registros de preços dos postos do _estado de São Paulo_ (True).

Para **filtrarmos** os registros de postos do estado de São Paulo:

O resultado é um Data Frame com _apenas_ os registros desejados após a **filtragem**.<br/>
Podemos ainda utilizar o método `loc` para o mesmo fim:

##### **==> Alternativa 2: Utilizando o método `query`**

`query` filtra linhas de um DataFrame baseado em uma **query (pergunta)**.

<br/>

Uma boa prática é **salvar o Data Frame filtrado em uma nova variável**. Isso simplifica a complexidade do código para futuras análise feita para os postos de São Paulo.

#### **Selecionando registros de postos do Rio de Janeiro com Preços acima de 2 reais**

Note que o resultado da seleção continua sendo uma _Series de booleans_ com o _mesmo número de linhas/amostras do DataFrame_, de modo que cada linha possuirá um valor booleano indicando se o posto é do Rio de Janeiro e o preço aferido do combustível é maior que 2 reais (True) ou não.

O símbolo **&** significa **AND** na comparação. Essa nomenclatura do python/pandas é diferente das nomenclaturas tradicionais (&&). <br/>
Similarmente:
- **|** representa o **OR** (não é ||)
- **~** representa o **NOT** (não é !)

Alternativamente, poderíamos usar o método `query` para fazermos tal seleção. Porém, isso não é possível especificamente para esse caso, pois o rótulo da coluna 'PREÇO MÉDIO REVENDA' possui caracteres inválidos para o método (cedilha, acentos) 

In [ ]:
# Não funciona
# data.query('ESTADO=="RIO DE JANEIRO" and PREÇO MÉDIO REVENDA > 2')

**Aprofundando mais ainda**

A primeira comparação `(data['ESTADO'] == 'RIO DE JANEIRO')` checa, linha a linha (amostra a amostra) do DataFrame, quais são aquelas cujo o estado é RIO DE JANEIRO. Nenhuma averiguação de preços é feita nesse momento. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

A segunda comparação `(data['PREÇO MÉDIO REVENDA'] > 2)` checa, linha a linha (amostra a amostra) do DataFrame, quais são os registro cujo preço do combustível é maior que 2 reais. Note que essa comparação checará os postos de **TODOS os estados**. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

Por fim, as duas "perguntas" são unidas pelo AND (&) que retorna a "pergunta completa" que fizemos.

Alguns podem argumentar que tal abordagem é **ineficiente**, uma vez que, para cada condição ("pergunta"), estamos varrendo todas as linhas do DataFrame. <br/>
O Pandas _tenta otimizar_ isso ao máximo por de trás dos panos. Mas, de fato, de tivermos um dataset **muito grande** (centenas de milhares de linhas), tal abordagem se tornará _lenta_.

Assim, poderíamos fazer filtragem com múltiplos condicionais em partes:

#### **Selecionando registros de postos de São Paulo ou do Rio de Janeiro com Gasolina Comum acima de 2 reais**

Podemos fazer a solução do "jeito mais lento", percorrendo o DataFrame inteiro _múltiplas vezes_:

<br/>

Alternativamente:

#### **Selecionando registros dos anos de 2008, 2010 e 2012**

**ALTERNATIVA 1**

**ALTERNATIVA 2**

### **Iterando com DataFrames**

#### For-each `DataFrame.iterrows()` (LENTO ==> apenas indicado para iterar pequenos conjunto de dados)

<h2>2. Preparação dos dados</h2>
<hr/>

### 2.1 Removendo amostras com valores vazios (null / nan) no dataset

De um total de 106823 observações, **não há valores null** / nan para nenhum atributo. Mas, veremos que não é bem assim neste caso específico.<br/><br/>

### 2.2 Conversão de tipos de atributos

O pandas automaticamente reconhece os tipos de dados de cada coluna. <br/>
Porém, existem alguns atributos que estão com seus tipos errados: P. ex., "PREÇO MÉDIO DISTRIBUIÇÃO" deveria ser ```float64``` e não ```object```.<br/>
Nestes casos, muito provavelmente algumas amostras têm um string ao invés de um número para tais atributos. <br/>

Os atributos *"DATA INICIAL"* e *"DATA FINAL"* deveriam ser do tipo `datetime`.

Em outros casos, alguns **atributos categóricos** são ```objects```, mas poderiam ter o tipo ```category```, que é um tipo especial do pandas. <br/>
Este tipo é necessário para se utilizar algumas funções específicas do pandas. <br/>
**Não** converteremos para este tipo por ora.

#### **Datas**
Como os atributos de data do datset já estão em um formato de data aceitável (YYYY-MM-DD), não precisamos forçar nenhuma conversão nesse sentido.

#### **Dados Numéricos**

<br/>

Note que temos vários valores ***null*** agora **após a *conversão de tipos***. Vamos checá-los com mais cuidado nos dados originais e preprocessados.

### 2.3 Limpeza de dados

Várias amostras possuem a _string '-'_ em algumas colunas ao invés de um número de fato. Ou seja, não há aferições destes atributos para estas amostras. <br/>

<br/>

Poderíamos **preencher os valores NaN com um valor padrão**. Para isso, basta usar o método `.fillna`.

<br/>

Por mais que a função `fillna` seja interessante e útil em muitos casos, no problema em questão estamos interessados em analisar precisamente, p. ex., o **'PREÇO MÉDIO DISTRIBUIÇÃO'**.<br/>
A fim de não termos valores _sintéticos_ gerados pelo `fillna`, que possam atrapalhar nossa análise, iremos **remover (drop) todas as amostras que possuem qualquer valor NaN** para quaisquer atributos/colunas. <br/>

Para isso, basta utilizarmos o método `dropna`.

Nosso data frame agora, após essa _limpeza_, ficou mais enxuto, contentdo 103392 registros frente aos 106823 registros originais. <br/>

Essas são apenas algumas das possíveis _técnicas de limpeza de dados_. Outras estratégias, p. ex., **confiam na detecção de outliers**, que não veremos neste curso.

#### **Salvando o Dataset Preprocessado**

<h2>3. Estatísticas Descritivas</h2>
<hr/>

O Pandas fornecem algumas funções/métodos ue computam certas estatísticas descritivas.

`describe`: exibe várias **estatísticas descritivas** para os _atributos_ de um dataframe ou para uma _Series_.

<br/>

Como o resultado do `describe` de um _dataframe_ é outro _dataframe_, podemos filtrar apenas algumas colunas.

**Acessando apenas algumas estatísticas**

<br/>

`mean`, `std`, `min`, etc: cada uma das estatísticas do `describe` podem ser computadas individualmente:

#### Qual é o menor preço mínimo de revenda?

#### Qual é a média e desvio padrão dos preços mínimos de revenda?

#### Quais são os estados considerados?

#### Quantos registros (aferições) cada estado possui?

`.value_counts()`:  Conta a frequência dos valores de uma dada variável (de preferência, _categórica_).

<h2>4. Executando funções para cada item de um DataFrame ou Series</h2>
<hr/>

Uma alternativa ao `for-loop` que vimos anteriormente e que é _lento_, é usarmos _funções próprias do pandas_ que **aplicam/mapeiam uma dada função a todos os elementos de um DataFrame ou Series**, retornando novos elementos "transformados".


<img src='./imagens/apply_map_applymap.png' width=300/>


Fonte: https://towardsdatascience.com/introduction-to-pandas-apply-applymap-and-map-5d3e044e93ff

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4], 
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]}, 
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])

In [ ]:
df

`apply()`: usado para aplicar uma função ao longo de um eixo de um DataFrame ou em valores de uma Series.

<img src='./imagens/pandas_axis.jpg' width=500/>

Fonte: https://www.allthesnippets.com/browse/pandas/df_axis.html

<img src='./imagens/apply_axis_1.png' width=250/>

<img src='./imagens/apply_axis_0.png' width=250/>

##### Usando `lambda` functions

<img src='./imagens/apply_axis_1_mean.png' width=350/>

<br/>

`applymap()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de um DataFrame.

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4], 
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]}, 
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])
df

<br/>

`map()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de uma _Series_.

<h2>5. Agrupamento</h2>
<hr/>

`groupby`: Usado para criar **grupo de elementos** (e.x., baseado nos valores de um atributo). <br/>
**Funções** podem então ser aplicadas para os _elementos de cada grupo_, de modo que os **resultados de cada grupo são combinados**.

<br/>

Também podemos ter agrupamentos por mais de um atributo.

<br/>

`.agg`: **agrega (roda)** uma série de funções para os elementos de um dataframe ou de grupos de um dataframe.

In [ ]:
df = pd.DataFrame([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9],
                   [np.nan, np.nan, np.nan]],
                  columns=['A', 'B', 'C'])
df

<h2>6. Ordenação</h2>
<hr/>

In [ ]:
notas = pd.DataFrame({
    'nome': ['João', 'Maria', 'José', 'Alice'],
    'idade': [20, 21, 19, 20],
    'nota_final': [5.0, 10.0, 6.0, 10.0]
})
notas

`.sort_values()`: ordena valores ao longo de um eixo.

Por padrão, o método retorna uma cópia dos dados ordenados em **ordem crescente (ascendente)**. Podemos alterar isso pelo argumento `ascending`.

<br/>

Podemos ordenar a partir de **mais de uma coluna**:

Ordena os registros, primeiramente, pela coluna 'nota_final' em **ordem descrente**. <br/>
Então, reordena os registros _"empatados"_, ou seja, com a **mesma nota final**, em _ordem alfabética_ (ordem crescente).

<br/>

Note que o dataframe original **não foi alterado** após a ordenação.

Para alterá-lo, use o argumento `inplace=True`:

<h2>7. Exercícios</h2>
<hr/>

Vamos aplicar os conceitos que vimos em alguns exercícios. <br/>
Para isso, utilizaremos o dataset de _preços de combustíveis no Brasil_.

Como há apenas medições de janeiro a junho para o ano de 2019, resolvemos **remover os dados** deste ano da análise.

Note que temos um novo dataframe com 99739 linhas, mas com índices fora desse intervalo. <br/>
Acontece que os registros mantiveram seus índices originais após a query. <br/>

Para resetar os índices de _0 a num_linhas-1_, basta usarmos o método `.reset_index()`. 

Os índices agora foram **resetados**. Porém, os índices antigos se transformaram em _uma nova coluna_ chamada 'index'. <br/>
Para removê-la durante o _reset_, basta passarmos o argumento `drop=True`.

### 7.1 Qual a proporção de postos pesquisados para cada combustível em cada região

### 7.2 Como os preços da Gasolina Comum em São Paulo variaram em 2018?

#### **Estatísticas Descritivas**

### 7.3 Como os preços da Gasolina Comum e do Etanol em São Paulo variaram em 2018?

<h2>8. Assuntos para continuar os estudos</h2>
<hr/>

- join
- concat
- plot
- data cleaning